promql/functions.go

```diff
@@ -716,16 +716,28 @@ func linearRegression(samples []Point, interceptTime int64) (slope, intercept fl
 		sumY, cY   float64
 		sumXY, cXY float64
 		sumX2, cX2 float64
+		initY      float64
+		constY     bool
 	)
-	for _, sample := range samples {
+	initY = samples[0].V
+	constY = true
+	for i, sample := range samples {
+		if constY && i > 0 && sample.V != initY {
+			constY = false
+		}
 		n += 1.0
 		x := float64(sample.T-interceptTime) / 1e3
 		sumX, cX = kahanSumInc(x, sumX, cX)
 		sumY, cY = kahanSumInc(sample.V, sumY, cY)
 		sumXY, cXY = kahanSumInc(x*sample.V, sumXY, cXY)
 		sumX2, cX2 = kahanSumInc(x*x, sumX2, cX2)
 	}
-
+	if constY {
+		if math.IsNaN(initY) || math.IsInf(initY, 0) {
```

Comment:
At this point, `math.IsNaN(initY)` will never be true because NaN never equals anything, including other NaN.

Therefore, I think we don't have to handle NaN. And for the case of ±Inf, I think the code is already doing the right thing, namely `return 0, initY` (if all Ys are plus or minus infinity, the intercept is also plus or minus infinity, and the slope is 0).

In short: I believe we don't need the special handling of NaN and Inf. `if constY { return 0, initY }` is doing the right thing.

Reply:
`rate(vector(1/0)[5m:])` is NaN, not +Inf (initY).

Reply:
You mean `deriv(vector(1/0)[5m:])`, don't you?

Which also evaluates to NaN currently, not 0, as I would have expected.

Reply:
Yes, deriv too. I expect that.

Reply:
updated in last commit